### PR TITLE
Add Mul type class

### LINKED ISSUE #1581: Add `Mul` Type Class
- [ ] Add the `Mul` type class.
- [ ] Add instances for `Float32`.
- [ ] Add instances for `Float64`.
- [ ] Add instances for `Int8`.
- [ ] Add instances for `Int16`.
- [ ] Add instances for `Int32`.
- [ ] Add instances for `Int64`.
- [ ] Add instances for `BigInt`.
- [ ] Rewrite `*` to be syntactic sugar for `Mul.mul`. 

Question: How should these instances be added? Compiler magic?

```
class Mul[a] {
  def mul(x: a, y: a): a
}
```

NB: I have arbitrarily chosen `Mul` as the first operator to try to typeclass-ify.

### PR DESCRIPTION
**Pull request by esbenbjerre**
- [x] Implement `Mul` type class.
- [x] Add instances for `Float32`, `Float64`, `Int8`, `Int16`, `Int32`, `Int64`, `BigInt`.
- [x] Add tests.
- [x] Rewrite `*` to be syntactic sugar for `Mul.mul`.

Fixes #1581.

### CODE REVIEW

**Comment by esbenbjerre** [comment]
I'm not sure where the refactoring is to be made in relation the `*` operator.

**Comment by magnus-madsen** [comment]
> 
> 
> I'm not sure where the refactoring is to be made in relation the `*` operator.

I think I can do that after.

**Comment by magnus-madsen** [comment]
I will write some instructions and then you can use them in the PR. Give me a few minutes :)

**Comment by magnus-madsen** [comment]
I think you should try the following:

```
          case "*" =>
            val qname = Name.mkQName("Mul.mul", sp1, sp2)
            val target = WeededAst.Expression.DefOrSig(qname, loc)
            WeededAst.Expression.Apply(target, List(e1, e2), loc)
            WeededAst.Expression.BinaryDeprecated(BinaryOperator.Times, e1, e2, loc)
```

here:

image

In the Weeder

**Comment by magnus-madsen** [comment]
If that does not seem to work (e.g. type issue), maybe we should use the `mkApplyFqn` helper in Weeder.

**Comment by esbenbjerre** [comment]
> We should probably just use `mkApplyFqn`.

Neat! It seems to work.
A highly scientific test where I remove the instance for `Int32` gives me a lot of complaints about a missing `Mul` instance.

Do you think we should use `*` or `Mul.mul` in tests?

**Comment by magnus-madsen** [comment]
> Do you think we should use `*` or `Mul.mul` in tests?

What do you think? I am leaning towards `*`.

File: main/test/ca/uwaterloo/flix/library/TestBigInt.flix [hunk 1]
```diff
@@ -745,4 +747,26 @@ namespace TestBigInt {
     @test
     def clampToFloat6407(): Bool = BigInt.clampToFloat64(-8982741234134123419879712341ii, -100.0f64, 100.0f64) == -100.0f64
 
+    /////////////////////////////////////////////////////////////////////////////
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestBigInt.flix:750]: I think these being tests for the basic operations-- we should just move them to the top of the test file. So its order goes:

- tests of primitives (in the order of SemanticOperator)
- tests of whatever is in e.g. Int32
-

**Comment by esbenbjerre** [comment]
> What do you think? I am leaning towards `*`.

Same. The only thing I wonder is whether there should be a test or two to make `*` is translated correctly. I'm not sure if that's trivial or not.

**Comment by magnus-madsen** [comment]
I think we need to decide what to do about the tests in `Test/Exp/Binary/Arithmetic`.

The reason they fail is because in `LangSuite` they are executed without the standard library.

**Comment by magnus-madsen** [comment]
OK; in the interest of not overburdening us, let us simply change line 80 in TestLang from:

```
  new FlixTest("Test.Exp.Binary.Arithmetic", "main/test/flix/Test.Exp.Binary.Arithmetic.flix"),
```

to

```
  new FlixTest("Test.Exp.Binary.Arithmetic", "main/test/flix/Test.Exp.Binary.Arithmetic.flix")(Options.TestWithLibrary)
```

Then I suggest that we later revisit how to structure the tests. We can keep the mul tests (make them use `*` and move them to the top), but we should not add new tests for e.g. `Add` before we decide on a more general testing strategy. There is already a lot of work, so let us just focus on getting the type classes implemented and reusing the existing tests.

**Comment by esbenbjerre** [comment]
Gotcha. I will have to do the same for `Test.Exp.Infix` as there are two tests failing in there as well.

**Comment by magnus-madsen** [comment]
Ready for merge? Then I suggest we do `Add`, `Sub`, and `Div` next, and so forth by chunks!

**Comment by esbenbjerre** [comment]
> Ready for merge? Then I suggest we do `Add`, `Sub`, and `Div` next, and so forth by chunks!

Yes I think it's ready.
Just to be sure, does "by chunks" refer to one at a time or as in one go?

**Comment by magnus-madsen** [comment]
> Just to be sure, does "by chunks" refer to one at a time or as in one go?

I mean, let us do: 3-4 at a time.

**Comment by magnus-madsen** [comment]
Just a ping to say that I updated my post above.

**Comment by esbenbjerre** [comment]
> Just a ping to say that I updated my post above.

Thanks, noted!

**Pull request merged**